src/olympia/addons/views.py

```diff
@@ -362,6 +363,32 @@ def destroy(self, request, *args, **kwargs):
 
         return super().destroy(request, *args, **kwargs)
 
+    def update(self, request, *args, **kwargs):
+        if kwargs.get('partial'):
+            # PATCH uses the standard update always
+            return super().update(request, *args, **kwargs)
+
+        # for PUT we check if the object exists first
+        try:
+            instance = self.get_object()
+        except http.Http404:
+            instance = None
+
+        # We only support guid style ids for PUT requests
+        if self.lookup_field != 'guid':
+            raise http.Http404
+
+        if instance:
+            # if the add-on exists we can use the standard update
+            return super().update(request, *args, **kwargs)
+
+        # otherwise we create a new add-on
+        serializer = self.get_serializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.save(guid=self.kwargs['guid'])
+
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
```

Comment:
Couldn't you use `self.create(request, *args, **kwargs)` here instead ?

Reply:
And, related to that... wouldn't `self.action` be incorrect here with your current code ?

Reply:
It is possible to use `self.create` (and I almost did) but we'd need to override the `save` method to include `guid` (and either check the action in there so we only did it for PUT) so I didn't think it helped much.  I can though?

Reply:
I'm not following what you mean about `self.action` - can you expand on it and/or include an example scenario?

Reply:
When you reach this code, `self.action` is still `update`, but we're doing a creation

Reply:
> It is possible to use `self.create` (and I almost did) but we'd need to override the `save` method to include `guid` (and either check the action in there so we only did it for PUT) so I didn't think it helped much. I can though?

No, you're right, it's not worth it.

Reply:
https://github.com/mozilla/addons-server/pull/19342/commits/4445dd36723a9b9c65df34918f93c8464e8647bc